nvflare/edge/device/android/sdk/training/ETTrainer.kt

```diff
@@ -0,0 +1,472 @@
+package com.nvidia.nvflare.sdk.training
+
+import android.content.Context
+import android.util.Log
+import com.nvidia.nvflare.sdk.training.TrainingConfig
+
+import com.nvidia.nvflare.sdk.core.Dataset
+import org.pytorch.executorch.Tensor
+import org.pytorch.executorch.EValue
+import org.pytorch.executorch.TrainingModule
+import org.pytorch.executorch.SGD
+import com.facebook.soloader.nativeloader.NativeLoader
+import com.facebook.soloader.nativeloader.SystemDelegate
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import kotlin.jvm.Throws
+
+/**
+ * Android ExecuTorch trainer implementation that matches iOS functionality.
+ * Implements AutoCloseable for proper resource management.
+ * 
+ * Usage:
+ * ```
+ * ETTrainer(context, modelData, meta).use { trainer ->
+ *     val dataset = XORDataset("train")  // or CIFAR10Dataset(context)
+ *     val result = trainer.train(config, dataset, modelData)
+ * }
+ * ```
+ */
+class ETTrainer(
+    private val context: android.content.Context,
+    private val modelData: String, 
+    private val meta: Map<String, Any>
+) : AutoCloseable {
+    private val TAG = "ETTrainer"
+    private var tModule: TrainingModule? = null
+    private var isInitialized = false
+    
+    companion object {
+        private const val EXECUTORCH_HEADER_SIZE = 8
+        private const val EXECUTORCH_HEADER_PREFIX = "PAAAAEVU"
+        private const val BUFFER_SIZE = 4 * 1024  // 4KB buffer for file operations
+        private const val PROGRESS_LOG_INTERVAL = 500  // Log progress every 500 steps
+        private const val DEFAULT_MOMENTUM = 0.9f
+        private const val SGD_WEIGHT_DECAY = 0.0
+        private const val SGD_NESTEROV = true
+        private const val LOSS_TENSOR_INDEX = 0
+        private const val PREDICTIONS_TENSOR_INDEX = 1
+    }
+    
+
+    
+    private val artifactManager = TrainingArtifactManager(context, meta)
+
+    init {
+        setupArtifactDirectories()
+    }
+
+    /**
+     * Setup artifact directories.
+     */
+    private fun setupArtifactDirectories() {
+        // Artifact setup is now handled by TrainingArtifactManager
+        Log.d(TAG, "Artifact directories setup complete")
+    }
+
+    /**
+     * Create dataset internally based on training method (iOS-style).
+     */
+    private fun createDataset(method: String, config: TrainingConfig): com.nvidia.nvflare.sdk.core.Dataset {
+        return when (method) {
+            "xor" -> {
+                Log.d(TAG, "Creating XOR dataset for training")
+                com.nvidia.nvflare.app.data.XORDataset("train")
+            }
+            "cnn" -> {
+                Log.d(TAG, "Creating CIFAR-10 dataset for training")
+                com.nvidia.nvflare.app.data.CIFAR10Dataset(context)
+            }
+            else -> throw IllegalArgumentException("Unsupported training method: $method")
+        }
```

Comment:
the dataset needs to be created and then passed in, so users will write their own XORDataset and then do:


AndroidFlareRunner(data_source=xxx)

instead of we create them inside here